### ISSUE TITLE
[Quest] Updating the Tasks generation to on demand instead of all at once

### ISSUE THREAD

**Issue by dubdabasoduba** [opensrp/fhircore#1696]
**Enhancement context**
- We currently generate all tasks for a CarePlan.
- We would like to update the task generation to an on-demand cycle. This means a task is only generated after a pre-requisite is completed or cancelled. 

**Implementation** 
- [ ] Generate Tasks on demand instead of generating all the Tasks at once. 
    - [ ] We however want to make this configurable since some items would want the Tasks generated all at once. e.g If we choose to visualize the child immunization schedules using tasks then we would all the tasks for the schedule to be generated at once. 
- [ ] Make sure that the schedule defined on the CarePlan and StructureMap is adhered to. 
- [ ] We will stop the generation of the tasks once the `executionPeriod.end` is >= to the CarePlan `period.end` date. or whent he `activity.scheduledTiming.repeat.countmax` is reached.  
- [ ] Always generate the first tasks on the day of CarePlan creation/generation.
- [ ] Adapt the task's `executionPeriod.end` & `executionPeriod.start` dates to the schedule depending on when the previous task was closed. e.g For the monthly ANC visit schedule
     - [ ] If task 1 had a start date of 01/01/2022 and an end date of 31/01/2022. Then the task was closed on 09/02/202
     - [ ] Then the next task to be generated will have a  start date of 01/03/2022 and an end date of 31/03/2022 instead of the start date of 01/02/2022 and end date of 28/02/2022

**Acceptance criteria**
- [ ] Only generate a task after the previous one has the status `completed` or `cancelled`
- [ ] The Task should follow the defined schedule on the CarePlan.

**Comment by pld** [comment]
This makes sense as the high level plan. What's the technical implementation plan for making sure we generate the new tasks appropriately?

We don't need that know, but who ever is assigned this, please sketch that out in a comment here and request review before coding begins, going to add a blocked label as a reminder of that.

**Comment by dubdabasoduba** [comment]
@pld Makes sense. @maimoonak and I are to have a call to define how to approach it. We will update the Technical approach added here.

**Comment by pld** [comment]
@maimoonak can you add an implementation plan to this ticket? I think that will unblock it

**Comment by maimoonak** [comment]
Hi  @pld @dubdabasoduba . Sorry for late response. I was exploring the current implementation of $apply. Looks like the changes are major and also we would need to
- Write CQL for all plan definitions
- Remove the StructureMap transforms to handle multiple actions and find a way to map CQL and the transform
- Update the current code to call $apply instead of fhir-task-generator

@dubdabasoduba mentioned that the requirement is urgent so we need to have a solid implementation plan rather than the major changes

For current implementation we can write $apply CQL in parallel and add the feature as a separate way of creating CarePlan.
- Keep the implementation same and update PlanDefinition and remove the repeat to 1 where we want to keep 1 task per Event. This would allow implementer configure the 'all-at-once' tasks or 'one-per-visit' tasks.
- Stopping the generation of the tasks once the executionPeriod.end is >= to the CarePlan period.end date is in place can be added as a separate condition. Rightnow we are marking CarePlan completed based on a specific type of Condition i.e. Sick Child got 'fully-recovered' into questionnaire.answer
- Adapt the task's executionPeriod.end & executionPeriod.start dates to the schedule would be done by passing the last task via 'input' -> DataRequirement into StructureMap. And the structure-map would handle the generation of new task. Plz note that currently we have updating the current careplan code inplace so it would be another action into PlanDefinition only.

**Comment by maimoonak** [comment]
the PR #1854  allows using both approaches.. i.e. $lite, $apply..The $ apply (2.1.0) can not be used with repeat timings... i.e. can not generate ALL tasks at once with repeat schedule. rather we need to specifiy each as different action. However, $lite can be used for both.

**Comment by pld** [comment]
OK cool that's reasonable, thanks, I doubt we'd want to use $apply unless we've already switch to on-demand generation.
